error-issue/src/stackdriver_api.ts

```diff
@@ -81,15 +84,25 @@ export class StackdriverApi {
     groupId?: string;
     'serviceFilter.service'?: string;
   }): Promise<Array<Stackdriver.ErrorGroupStats>> {
-    const {errorGroupStats = []} = (await this.request('groupStats', 'GET', {
-      'timeRange.period': 'PERIOD_1_DAY',
-      timedCountDuration: `${SECONDS_IN_DAY}s`,
-      ...opts,
-    })) as {errorGroupStats?: Array<Stackdriver.SerializedErrorGroupStats>};
+    const cacheKey = opts['serviceFilter.service'] || 'ALL_SERVICES';
+    if (!this.cache.has(cacheKey)) {
+      const {errorGroupStats = []} = (await this.request('groupStats', 'GET', {
+        'timeRange.period': 'PERIOD_1_DAY',
+        timedCountDuration: `${SECONDS_IN_DAY}s`,
+        ...opts,
+      })) as {errorGroupStats?: Array<Stackdriver.SerializedErrorGroupStats>};
 
-    return errorGroupStats.map((stats: Stackdriver.SerializedErrorGroupStats) =>
-      this.deserialize(stats)
-    );
+      this.cache.set(
+        cacheKey,
+        errorGroupStats.map((stats: Stackdriver.SerializedErrorGroupStats) =>
+          this.deserialize(stats)
+        )
+      );
+    } else {
+      console.info('Returning Stackdriver results from local cache');
+    }
+
+    return this.cache.get(cacheKey);
```

Comment:
The structure here is kind of weird - when dealing with caches assume little of its concurrency (even though node-cache probably uses in-process memory, but still...)

I would rearrange this as:
```ts
let errorGroupStats?: Array<Stackdriver.SerializedErrorHroupStats> = this.cache.get(cacheKey);
if (!errorGroupStats) {  // or "if(errorGroupStats === undefined)" if you're gonna store [] in the cache
  console.info('Cache miss for', cacheKey, '- fetching from Stackdriver');
  errorGroupStats = ... // fetch from stackdriver
  this.cache.set(cacheKey, errorGroupStats);
} else {
  console.info('Cache hit for', cacheKey);
}

return stackGroupStats || [];
```

(I'm sure I got a few things wrong in terms of APIs but I'm also sure you can fix them easily 🙃)

Reply:
Done